Support per-channel quantized INT8 weights unpacking in XNNPACK delegate

dev0x13 opened the pull request:
This MR expands INT8 weights unpacking for FP32 inference in XNNPACK delegate support added in [the previous MR](https://github.com/tensorflow/tensorflow/pull/48751) for per-channel dynamic range quantized model. Previous changes have only supported per-tensor quantization mode, which is obsolete in recent TensorFlow releases.
I have not added proper testing yet, because I'd like to get some suggestions from a maintainer, i.e. the best way to organize such testing while keeping the codebase clean (specifically, how to change testers). Thank you in advance!

dev0x13 committed:
Support per-channel quantized INT8 weights unpacking in XNNPACK delegate

dev0x13 commented:
@multiverse-tf Could you review this please?

dev0x13 committed:
Merge branch 'master' into xnnpack-per-channel-int8-weights-unpack

dev0x13 committed:
Mention quantized INT8 weights unpacking option in XNNPACK delegate README

gbaned commented:
@dev0x13  Can you please resolve conflicts? Thanks!

dev0x13 committed:
Merge branch 'master' into xnnpack-per-channel-int8-weights-unpack

dev0x13 commented:
@gbaned Done

dev0x13 commented:
@multiverse-tf Could you review this please?

dev0x13 commented:
It's been more than a month since I've created this MR. Should I close it now due to the lack of review?

multiverse-tf commented:
> It's been more than a month since I've created this MR. Should I close it now due to the lack of review?

Sorry for not catching this review request earlier. In the past month, I think we've also added native per-channel quantized INT8 op support in XNNPACK. @Maratyszcza, could you shed more light on the support? Thx

dev0x13 commented:
Closing due to the lack of response from maintainers.

dev0x13 committed:
Merge branch 'master' into xnnpack-per-channel-int8-weights-unpack

multiverse-tf commented:
@Maratyszcza could you help with this review? Thx!

dev0x13 committed:
Merge branch 'master' into xnnpack-per-channel-int8-weights-unpack

dev0x13 committed:
Add tests for per-channel dynamic-range INT8 quantization for XNNPACK delegate

dev0x13 committed:
Adjust XNNPACK delegate README section on dynamic-range qunatization  

Dynamic-range quantization is enabled by default now, so the special Bazel build flag mention is no longer needed.

dev0x13 committed:
Merge branch 'master' into xnnpack-per-channel-int8-weights-unpack

dev0x13 commented:
@multiverse-tf @Maratyszcza Could you review this please? Thanks in advance!

dev0x13 commented:
@multiverse-tf @Maratyszcza Could you review this please? This MR is ready to be merged, only review is needed.

Maratyszcza commented on tensorflow/lite/delegates/xnnpack/README.md:
```diff
@@ -342,7 +342,7 @@ must be enabled by adding extra Bazel flags when building TensorFlow Lite.
 operators using signed quantization schema. This schema is used by models
 produced by [Model Optimization Toolkit](https://www.tensorflow.org/model_optimization)
 through either post-training integer quantization or quantization-aware
-training. Post-training dynamic range quantization is not supported in XNNPACK.
```
This is not dynamic range quantization. Models with dynamic range quantization quantized floating-point inputs to Conv2D/FC/etc, do quantized computations, then dequantize the outputs.

dev0x13 replied:
I am not quite sure about that. From my experience I have concluded that dynamic-range quantization implies that weights are quantized for storage, but at runtime almost all of them (not all because hybrid kernels are involved) are being dequantized and then all the computations are performed in floating-point. I got this understanding by inspecting converted TFLite models with Netron viewer. Some of my observations are described in [this](https://github.com/tensorflow/tensorflow/pull/48751#issue-867023983) message. Please correct me if I my conclusions are wrong.

Maratyszcza replied:
TFLite converter creates operators with `FP32` inputs/outputs and `INT8` weights when using dynamic-range quantization. These operators are processed by so-called hybrid kernels, which dynamically quantize input and dequantize output. XNNPACK currently doesn't support hybrid schemas, so the models converted with dynamic range quantization will not on the XNNPACK delegate, and it would be confusing to state otherwise.

Maratyszcza commented on tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc:
```diff
@@ -193,7 +193,7 @@ std::vector<char> BinaryElementwiseTester::CreateTfLiteModel(
   std::vector<flatbuffers::Offset<OperatorCode>> operator_codes{
       {CreateOperatorCode(builder, binary_op)}};
-  if (FP16Weights() || INT8Weights()) {
+  if (FP16Weights() || (INT8Weights() || INT8ChannelWiseWeights())) {
```
Extra parenthesis not needed

dev0x13 replied:
Fixed

Maratyszcza commented on tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc:
```diff
@@ -223,11 +225,33 @@ std::vector<char> BinaryElementwiseTester::CreateTfLiteModel(
+            static_cast<int32_t>(Input1Shape().size()) - 1;
+        const int32_t num_scales = Input1Shape()[input1_quantized_dimension];
+        input1_zero_points.resize(num_scales, 0);
+        auto scale_rng = std::bind(
```
Compute per-channel initialization parameters at runtime.

dev0x13 replied:
Done

Maratyszcza commented on tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc:
```diff
@@ -246,7 +270,9 @@ std::vector<char> BinaryElementwiseTester::CreateTfLiteModel(
-  float input2_scale = 0;
+  std::vector<float> input2_scales;
+  std::vector<int64_t> input2_zero_points;
+  int32_t input2_quantized_dimension;
```
Initialize to 0, as this is the default value in `CreateQuantizationParameters`

dev0x13 replied:
Fixed

Maratyszcza commented on tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc:
```diff
@@ -264,11 +290,33 @@ std::vector<char> BinaryElementwiseTester::CreateTfLiteModel(
-            input2_data.begin(), input2_data.end(),
-            quantized_input2_data.begin(),
-            std::bind(QuantizeInt8, std::placeholders::_1, 0, input2_scale));
+        input2_quantized_dimension = 0;
```
Combine with variable definition

dev0x13 replied:
Fixed

Maratyszcza reviewed (changes requested):
Thank you for the contribution, and apologies for the delays with the review. The PR needs some changes:
1. Compute per-channel quantization parameters at runtime rather than random-initialize.
2. Add support for the newly added TransposeConv operator

dev0x13 committed:
Merge branch 'master' into xnnpack-per-channel-int8-weights-unpack

dev0x13 committed:
Support per-channel quantized INT8 weights unpacking in XNNPACK delegate (MR review fixes)

1. Minor cosmetic fixes suggested by MR reviewer are applied to XNNPACK
delegate test set.
2. Tests for per-channel quantized INT8 weights unpacking in XNNPACK
delegate are refactored to perform proper per-channel quantization
parameters computation at runtime instead of using random-initialized
quantized tensors.
3. Added per-tensor and per-channel quantized weights unpacking
tests for TRANSPOSE_CONV op in XNNPACK delegate.

dev0x13 commented:
@Maratyszcza Thank you for the review! All the changes are made.

Maratyszcza commented on tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc:
```diff
@@ -66,7 +66,7 @@ void BinaryElementwiseTester::Test(tflite::BuiltinOperator binary_op,
+  if (FP16Weights() || INT8Weights() || INT8ChannelWiseWeights()) {
     ASSERT_TRUE(Input1Static() || Input2Static());
   }
 
```
Check that if channelwise weights are used, the static input has at least one dimension (i.e. isn't a scalar)

dev0x13 replied:
Done

Maratyszcza replied:
Should be `if (INT8ChannelWiseWeights() && Input1Static())` ... + the same for input2

dev0x13 replied:
Sorry, my bad. Fixed

Maratyszcza commented on tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc:
```diff
@@ -223,11 +225,29 @@ std::vector<char> BinaryElementwiseTester::CreateTfLiteModel(
-            std::bind(QuantizeInt8, std::placeholders::_1, 0, input1_scale));
+        input1_scales.resize(1);
+        input1_zero_points.resize(1, 0);
+        input1_scales[0] = GetInt8QuantizationScale(input1_data);
```
Combine with resize

dev0x13 replied:
Done

Maratyszcza commented on tensorflow/lite/delegates/xnnpack/test_util.cc:
```diff
@@ -33,15 +34,79 @@ int8_t QuantizeInt8(float value, int32_t zero_point, float scale) {
+  std::vector<int> current_dim(num_dims, 0);
+
+  do {
+    size_t offset =
```
const size_t

dev0x13 replied:
Fixed

Maratyszcza commented on tensorflow/lite/delegates/xnnpack/test_util.cc:
```diff
@@ -33,15 +34,79 @@ int8_t QuantizeInt8(float value, int32_t zero_point, float scale) {
+                            current_dim.data(), 0, nullptr);
+    const int channel_idx = current_dim[quantized_dimension];
+    const float val = data[offset];
+    if (has_min_max_value[channel_idx]) {
```
Initialize min to `-std::numeric_limits<float>::infinity()` and max to `+std::numeric_limits<float>::infinity()` and remove these checks

dev0x13 replied:
Done

Maratyszcza reviewed (changes requested):
LG overall. Please revert the change in README and simplify the code per comments, and we're good to go.

dev0x13 committed:
Support per-channel quantized INT8 weights unpacking in XNNPACK delegate (MR review fixes)

1. Reverted changes in XNNPACK delegate README.
2. Made some adjustments for XNNPACK tests code suggested by the MR
   reviewer.

dev0x13 commented:
@Maratyszcza Thank you! All the suggested changes are made.

Maratyszcza commented on tensorflow/lite/delegates/xnnpack/test_util.cc:
```diff
@@ -33,15 +34,72 @@ int8_t QuantizeInt8(float value, int32_t zero_point, float scale) {
+                            current_dim.data(), 0, nullptr);
+    const int channel_idx = current_dim[quantized_dimension];
+    const float val = data[offset];
+    if (min[channel_idx] > val) {
```
These lines can be further simplified using `std::min`, `std::max`

dev0x13 replied:
Done

dev0x13 committed:
Support per-channel quantized INT8 weights unpacking in XNNPACK delegate (MR review fixes)

Maratyszcza commented:
LGTM. @gbaned could you merge?

dev0x13 commented:
I noticed that the PR was rolled back right after merge. Is there something wrong with it?

Maratyszcza commented:
`transpose_conv_test` failed under sanitizers with `UndefinedBehaviorSanitizer: float-cast-overflow tensorflow/lite/delegates/xnnpack/test_util.cc:32:28`

dev0x13 commented:
@Maratyszcza Fixed this in [my fork](https://github.com/dev0x13/tensorflow/commit/bc81bf3c74c47bb07aea6c2d2dbb1887d83d71a9), but it seems that this PR cannot be reopened. What do I need to do in order to submit this fix to make my changes merged again? Thank you in advance!

Maratyszcza commented:
`prelu_test` fails under asan too, with the same error

dev0x13 commented:
@Maratyszcza [Fixed](https://github.com/dev0x13/tensorflow/commit/efa0a06559b2f980fc967a9dde8338997aadb8ac)

Maratyszcza commented:
Re-landed in a6d352f031f272972da0e6271de07e8b40cac74a

dev0x13 commented:
Thank you!
